torch/csrc/jit/passes/freeze_module.cpp

```diff
@@ -0,0 +1,307 @@
+  //       attributes {
+  //          B = <SubModule2 at ...>
+  //       }
+  //       submodules {
+  //         module SubModule2 {
+  //            attributes {
+  //               scale = 2
+  //            }
+  //         }
+  //       }
+  //     }
+  //   }
+  //
+  // findConstantAttr(%B, "scale", M)  returns true because there are no
+  // explicit SetAttr that modifies %B. attrModule points to the module where
+  // attribute lives (in this example it is <SubModule2 at ...>).
+  //
+  // Note inplace mutations to attributes are checked later using alias
+  // analysis.
+  //
+  // We can use a more efficient algorithm to hash each constant GetAttr to its
+  // corresponding value. Based on initial test on resnet50 and other torch
+  // vision tests. GetAttrs are not too frequent so it is ok to chase GetAttr
+  // chain to retrieve their values.
+  bool findConstantAttr(
+      Node* node,
+      std::string& name,
+      script::Module& attrModule) {
+    std::stack<std::string> names;
+    while (!(node->outputs()[0]->type() == attrModule.type())) {
+      if (node->kind() == prim::GetAttr) {
+        names.push(node->s(attr::name));
+        node = node->inputs()[0]->node();
+      } else {
+        return false;
+      }
+    }
+
+    while (!names.empty()) {
+      auto moduleName = names.top();
+      names.pop();
+      attrModule = attrModule.attr(moduleName).toModule();
+      auto it = preservedAttrs_.find(attrModule._ivalue());
+      if (it != preservedAttrs_.end()) {
+        if (it->second.count(moduleName)) {
+          return false;
+        }
+      }
+    }
+    auto it = preservedAttrs_.find(attrModule._ivalue());
+    return it == preservedAttrs_.end() || !it->second.count(name);
+  }
+
+  void recordMutableAttrs(std::shared_ptr<Graph>& graph) {
+    std::stack<Block*> blocks({graph->block()});
+    std::unique_ptr<AliasDb> aliasDb =
+        torch::make_unique<AliasDb>(graph, /* isFrozen */ true);
+    while (!blocks.empty()) {
+      Block* block = blocks.top();
+      blocks.pop();
+      for (auto n : block->nodes()) {
+        for (Block* sub_block : n->blocks()) {
+          blocks.push(sub_block);
+        }
+        if (n->kind() == prim::SetAttr || n->kind() == prim::GetAttr) {
+          auto name = n->s(attr::name);
+          auto inputNode = n->inputs()[0]->node();
+          auto attrModule = module_;
+          if (!findConstantAttr(inputNode, name, attrModule)) {
+            continue;
+          }
+          if (n->kind() == prim::SetAttr || aliasDb->hasOutputWriters(n)) {
+            GRAPH_DEBUG(
+                n->kind() == prim::GetAttr ? "attribute: " + name + " in %" +
+                        n->outputs()[0]->debugName() + " has inplace writer"
+                                           : "");
+            preservedAttrs_[attrModule._ivalue()].insert(name);
+          }
+        }
+      }
+    }
+  }
+
+  std::set<std::string> recordReferencedAttrs(std::shared_ptr<Graph>& graph) {
```

Comment:
We do need it because we need to preserve the top level sub modules that contains mutable attributes. In the first recording step, we record only leaf attributes (for example  in example below only attribute"b" as mutable). In the second step, we record attributes that contain mutable attributes (such as "sub" attribute)  to preserve them.  

class Submodule (...)
      def __init_
           self.a = 
           self.b = 
      def forward(self, ...):
           self.b +=1
           return self.a